src/cir.rs

```diff
@@ -0,0 +1,451 @@
+// cir:  compact intermediate representation
+use crate::ast::Predicate;
+use crate::context::{Context, Match};
+use crate::interpreter::Execute;
+use crate::lir::{is_operator, LirInstruction, LirLogicalOperator, LirProgram, Translate};
+use crate::semantics::FieldCounter;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct CirProgram {
+    pub(crate) instructions: Vec<CirInstruction>,
+}
+
+impl CirProgram {
+    pub fn new() -> Self {
+        Self {
+            instructions: Vec::new(),
+        }
+    }
+}
+
+impl Default for CirProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum CirInstruction {
+    AndIns(AndIns),
+    OrIns(OrIns),
+    NotIns(NotIns),
+    Predicate(Predicate),
+}
+
+#[derive(Debug)]
+pub struct AndIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct OrIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct NotIns {
+    right: CirOperand,
+}
+
+#[derive(Debug, Clone)]
+pub enum CirOperand {
+    Index(usize),
+    Predicate(Predicate),
+}
+
+impl CirOperand {
+    #[inline]
+    pub fn as_predicate(&self) -> &Predicate {
+        match &self {
+            CirOperand::Index(_index) => {
+                panic!("Call as_predicate on index Operand, CirProgram is wrong.")
```

Comment:
```suggestion
                panic!("unexpected call to as_predicate with index Operand.")
```

(The error message should start with lower case.)

And the same for the following.

Reply:
fixed, thanks